python_modules/dagster/dagster/_core/workspace/context.py

```diff
@@ -501,15 +512,47 @@ def rm_state_subscriber(self, token: int) -> None:
         if token in self._state_subscribers:
             del self._state_subscribers[token]
 
+    def _reload_location_from_origin(self, origin: CodeLocationOrigin) -> Optional[CodeLocation]:
+        if isinstance(origin, ManagedGrpcPythonEnvCodeLocationOrigin):
+            endpoint = self._grpc_server_registry.reload_grpc_endpoint(origin)
+
+            return GrpcServerCodeLocation(
+                origin=origin,
+                server_id=endpoint.server_id,
+                port=endpoint.port,
+                socket=endpoint.socket,
+                host=endpoint.host,
+                heartbeat=True,
+                watch_server=False,
+                grpc_server_registry=self._grpc_server_registry,
+            )
+        elif isinstance(origin, GrpcServerCodeLocationOrigin):
+            # Disable the watcher thread while we reload
+            location_name = origin.location_name
+
+            with self._lock:
+                event = self._watch_thread_shutdown_events[location_name]
+                thread = self._watch_threads[location_name]
+
+                del self._watch_thread_shutdown_events[location_name]
+                del self._watch_threads[location_name]
+
+            if event:
+                event.set()
+                thread.join()
+
+            try:
+                new_location = origin.reload_location()
+            finally:
+                self._start_watch_thread(origin)
+
+            return new_location
```

Comment:
This is super gross and the main thing keeping me from pushing this PR forward right now. The code here is not particularly correct, but the problem that it's intending to solve is:

- User initiates a reload from a graphql request, that happens synchronously during the request
- There are also these background threads that watch for updates to the gRPC server and prompt you to update if the server ID changes, they will pick up on the change
- We don't want that - the user is initiating the request, so there's no need to pop something up telling them to reload

So this was intended to keep that from happening - can't come up with a solution I like to deal with that

Reply:
Other users observing the same dagit instance *should* get the prompt to reload their stale copy of the workspace when this fires though right? If the only problem is the dagit ux, it may require a dagit solution of ignoring/disabling the subscription when the mutation is firing.

Reply:
oh hm, that is a good point. I kinda feel like the whole experience around that prompt is not so good, and fixing that might resolve this - why doesn't it just update things in place when it detects there has been a change?

Reply:
i think its to avoid the risk of blowing things away when a user "is in the middle of something" in the not so happy paths like where whole jobs are disappearing  / moving

Reply:
OK yeah fair enough - pursuing the frontend 'disable the prompt while this mutation is happening' angle with @hellendag , will see how hard that is